Implement KEP-2535 - Ensure secret pulled images feature

sairameshv opened the pull request:
Developed these code changes on top of https://github.com/kubernetes/kubernetes/pull/114847 and https://github.com/kubernetes/kubernetes/pull/94899 
/cc @pacoxu @mikebrow 

Based on the latest proposal of the ensure secret pulled image feature, two fields in the kubelet configuration have been defined
Reference: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2535-ensure-secret-pulled-images#proposal

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
Extends the existing kubelet configuration with two new fields according to the KEP https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2535-ensure-secret-pulled-images#proposal
<!--
#### Which issue(s) this PR fixes:

*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*

-->
Fixes https://github.com/kubernetes/enhancements/issues/2535
#### Special notes for your reviewer:
As of today, if a pod pulls an image from a private repository with proper credentials, the image gets pulled onto a node successfully. When an another pod on the same node with wrong credentials tries to pull the same image, the container is getting started without an issue which shouldn't be the case. The above mentioned feature in the KEP helps in resolving this issue.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2535-ensure-secret-pulled-images/README.md
<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2535-ensure-secret-pulled-images/README.md 
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
NONE
```

mikebrow committed:
ensure secret pulled images; add PullImageSecretRecheckPeriod to option in

Signed-off-by: Mike Brown <brownwm@us.ibm.com>
Signed-off-by: Paco Xu <paco.xu@daocloud.io>
Signed-off-by: Sai Ramesh Vanka <svanka@redhat.com>

pacoxu committed:
set PullImageSecretRecheckPeriod default to 0; default not recheck for IfNotPresent Image

pacoxu committed:
if pullImageSecretRecheckPeriod>0 and KubeletEnsureSecretPulledImages is enabled, do recheck logic
- if recheck is enabeld, recheck no auth image(pullCredentialsHash=="") as well

sairameshv commented:
/cc @haircommander

aojea commented:
let's add APIs and functionality and test together please, or we may end in inconsistent states

sairameshv commented:
> let's add APIs and functionality and test together please, or we may end in inconsistent states

Ack, Sure @aojea !

sairameshv committed:
Update and define new API fields in the Kubelet Configuration

Based on the proposal of the ensure secret pulled image feature,
two fields in the kubelet configuration have to be defined
Reference: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2535-ensure-secret-pulled-images#proposal

Signed-off-by: Sai Ramesh Vanka <svanka@redhat.com>

sairameshv commented:
/test all

haircommander commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -64,27 +87,45 @@ func NewImageManager(recorder record.EventRecorder, imageService kubecontainer.I
-	if container.ImagePullPolicy == v1.PullNever {
+func shouldPullImage(container *v1.Container, imagePresent, pulledBySecret, ensuredBySecret, pullImageSecretRecheck bool) bool {
+	switch container.ImagePullPolicy {
+	case v1.PullNever:
```
we still need to do some checking here. if it's not in the cache then we need to fail because the image may be present on the node but we're not allowed to use it

haircommander replied:
ah nevermind, this function is just answering the question "should we pull this image" which the answer for pull never is always no

sairameshv replied:
This scenario has been taken care before calling the `shouldPullImage` i.e. in the `isEnsuredBySecret` function.

sairameshv commented:
/test pull-kubernetes-unit

sairameshv commented:
/test all

haircommander commented on pkg/features/kube_features.go:
```diff
@@ -369,6 +369,14 @@ const (
 
+	// owner: @mikebrow @pacoxu
+	// kep: http://kep.k8s.io/2535
+	// alpha: v1.30
```
this should be 1.31

haircommander commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -64,27 +87,48 @@ func NewImageManager(recorder record.EventRecorder, imageService kubecontainer.I
 	} else {
 		puller = newParallelImagePuller(imageService, maxParallelImagePulls)
 	}
+	var pullImageSecretCheck bool
+	if pullImageSecretRecheck != nil {
```
if we also check whether the feature gate is enabled here also, then we can always refer to pullImageSecretCheck instead of checking both

sairameshv replied:
Yes, True that. Addressed.

haircommander commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -164,17 +244,85 @@ func (m *imageManager) EnsureImageExists(ctx context.Context, pod *v1.Pod, conta
+func (m *imageManager) loadEnsureSecretPulledImagesData() {
+	_, err := os.Stat(imageStateManagerPath)
+	if err != nil && !os.IsExist(err) {
+		klog.ErrorS(err, "Failed to stat file", "file", imageStateManagerPath)
```
is this an error? couldn't it be expected? Maybe this should be InfoS....

haircommander commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -224,3 +372,97 @@ func applyDefaultImageTag(image string) (string, error) {
+// 2. ensuredBySecret: true if the secret for an auth used to pull an
+// image has already been authenticated through a successful pull request
+// and the same auth exists for this podSandbox/image.
+func (m *imageManager) isEnsuredBySecret(imageRef string, image kubecontainer.ImageSpec, pullSecrets []v1.Secret) (bool, bool, error) {
```
can you name the bools in the return so we can easier keep track of which is which?

sairameshv replied:
Sure, Addressed

sairameshv replied:
Oh okay, Got it !

haircommander commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -224,3 +372,97 @@ func applyDefaultImageTag(image string) (string, error) {
+	}
+	// if the pullImageSecretRecheckPeriod is 0, we don't need to check the secret again
+	if m.pullImageSecretRecheckPeriod.Duration == 0 {
+		return pulledBySecret, true, nil
```
I don't know if this is right. If the duration is 0, then we don't delete the entry in the cache in refreshCache.

If pulledBySecret is true, then it's present in the cache, but not necessarily having matching auths (you check the auths below)
if it's false, then it's either the first time we have pulled the image, or the image has had it's auths cleaned up from the cache, and we should recheck.

sairameshv replied:
Yeah, I agree with you here, I have already added that condition and this piece of code is from the previous implementation which I felt could be removed. Thanks for pointing out !

sairameshv commented:
/test all

sairameshv commented:
/label api-review

kwilczynski commented on pkg/kubelet/container/helpers.go:
```diff
@@ -126,6 +127,17 @@ func pickFieldsToHash(container *v1.Container) map[string]string {
 	return retval
 }
 
+// HashAuth - returns a hash code for a CRI pull image auth, and an error if the hash cannot be calculated.
```
A small nit: Go style comments are slightly different.

sairameshv replied:
It is explaining what it does and returns. What more can we add?

kwilczynski replied:
It's not about what it does, but how it does it. Look at some other Go function comments. There is a specific comment syle Go follows.

kwilczynski commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -50,12 +59,26 @@ type imageManager struct {
 	podPullingTimeRecorder ImagePodPullingTimeRecorder
+
+	keyring *credentialprovider.DockerKeyring
+	// ensureSecretPulledImages - map of imageref (image digest) to successful secret pulled image details
```
I am not sure if this comment adds any value. One can see the type of the map values here clearly.

sairameshv replied:
Yeah, on the other note, I felt extra detailing was good and hence did not delete the comment.

cici37 commented:
/remove-sig api-machinery

kwilczynski commented on pkg/kubelet/images/puller.go:
```diff
@@ -31,6 +31,8 @@ type pullResult struct {
 	imageSize    uint64
 	err          error
 	pullDuration time.Duration
+	// pullCredentialsHash of successful pull with auth, nil if no auth used or error
```
This comment also does not add a lot of value. Plus, you cannot set strings to nil in Go.

sairameshv replied:
Addressed

sairameshv commented:
/retest

haircommander commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -50,12 +59,26 @@ type imageManager struct {
 
 var _ ImageManager = &imageManager{}
 
+const (
+	imageStateManagerPath = "/var/lib/kubelet/image_state_manager"
```
nit: this can be one line

haircommander commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -224,3 +375,93 @@ func applyDefaultImageTag(image string) (string, error) {
+	}
+
+	// if the image is in the ensured secret pulled image list, it is pulled by secret
+	pulledBySecret = m.ensureSecretPulledImages[imageRef] != nil
```
Okay one remaining structural concern: how can the kubelet differentiate between image that was present at startup and image that was previously pulled without a secret? currently, if we're checking just this map, then we'd assume that all images currently present on the node (before the kubelet initialized the cache) would be pulled without auth--which may not be true. we similarly have to be careful about [deleting](https://github.com/kubernetes/kubernetes/pull/125817/files#diff-b22ab169cdc70b2e968bb653f29cf0a1384d3bccaebeef96aebf2a5a40821bbeR460) an entry from the map, as we don't want to accidentally report it as an image that has no auth.

I am thinking we need to store an entry if we successfully pull without auth, and use some signifier for such a situation in the ensureSecretPulledImages map (maybe an empty string in the Auths map)? 

WDYT @sairameshv

liggitt replied:
> Okay one remaining structural concern: how can the kubelet differentiate between image that was present at startup and image that was previously pulled without a secret?

This was exactly the issue with the attempt in 1.30, and what made the previous attempt slip. Can I get a summary of how the approach has changed to address that issue (which in my mind is the main problem to be solved with this feature).

distinguishing between these is important and we have to be clear about how we are doing it without regressing existing users who preload images to be used without auth to be robust against public registry unavailability:
- present at kubelet startup, not pulled with auth (preloaded, expected to be usable by all pods)
- present at kubelet startup, previously pulled with auth

haircommander replied:
> distinguishing between these is important and we have to be clear about how we are doing it without regressing existing users who preload images to be used without auth to be robust against public registry unavailability:
   >  present at kubelet startup, not pulled with auth (preloaded, expected to be usable by all pods)

I actually think we should not support this case.  A user can turn off this feature if they're sensitive to registry unavailability enough to prepopulate the node. I think we should always treat a cache miss the same: repull the image. I think in the future, we can investigate ways to expose image management as a higher level api (kubectl image pull...) or something to allow users to prepopulate a node before an upgrade without needing to run pods to have the kubelet be aware of the image. but for now, I don't think we should support this case with this feature. WDYT?

haircommander replied:
one difference between this implementation and the last is we reintroduced kubelet saving the cache, so we can actually solve the 
```
present at kubelet startup, previously pulled with auth
```
case, as well as we use PullImageSecretRecheckPeriod == 0 to say "never delete the cache", where it was previously used to set PullImageSecretRecheck=false. This means if an image was ever pulled by the kubelet, we'll continue to track it and be able to protect against registry flakes on repulls with the same auth

sairameshv replied:
>I am thinking we need to store an entry if we successfully pull without auth, and use some signifier for such a situation in the ensureSecretPulledImages map (maybe an empty string in the Auths map)?

Currently, we enforced a check to store the metadata of the images only if they have auth component [here](https://github.com/kubernetes/kubernetes/pull/125817/files#diff-b22ab169cdc70b2e968bb653f29cf0a1384d3bccaebeef96aebf2a5a40821bbeR259-R272) 

>present at kubelet startup, previously pulled with auth

I think we have a solution with this implementation here. We do get the Image reference [again](https://github.com/kubernetes/kubernetes/pull/125817/files#diff-b22ab169cdc70b2e968bb653f29cf0a1384d3bccaebeef96aebf2a5a40821bbeR259) of an existing image and check the auth component and store it into the cache while performing this `EnsureImageExists` check during the container start.
So, we don't have a problem if a new container is about to start using an existing image that is pre-populated on the node. Does that makes sense @haircommander ?

liggitt replied:
> I actually think we should not support this case. A user can turn off this feature if they're sensitive to registry unavailability enough to prepopulate the node.

I'm having a hard time understanding how this matches what you said later... it sounds like if we persist the cache of what the kubelet pulled with auth, and only enforce checking auth for those images, we get the behavior I'm looking for for pre-existing images the kubelet didn't pull with auth (registry availability isn't blocking).

haircommander replied:
I think we could get away with what you're describing, but it makes the writing/reading of the cache from disk more important. If the read fails or the file is corrupted, we'd treat the present images as "prepulled". I guess my point is the only way to ensure we always re-auth when we need to is have the internal representation of the cache be the source of truth, and distrust any image not present there.

haircommander replied:
if you're not worried about that case, then I am not feeling super stubborn about it. But I think it simplifies the logic if we hold that invariant true.

liggitt replied:
> I guess my point is the only way to ensure we always re-auth when we need to is have the internal representation of the cache be the source of truth, and distrust any image not present there.

How we transition existing users onto this feature seamlessly is one of the biggest considerations I have.

I don't see how distrusting pre-existing images on first kubelet startup with this new code could ever be compatible with current behavior, unless we *also* require explicit configuration of which pre-existing images the admin intends to be available without auth. That seems really fiddly / difficult to make the admin specify, and I hadn't seen any discussion of that, so I was assuming nothing like that was planned.

pacoxu replied:
1. now, users are not enable this 
2. for instance, in v1.31, we add the FG, and we enable the cache by default, but not use it. FG is alpha and disabled by default. But the cache saving is enabled.
3. in next releases(if we support upgrading node from 1.31 to 1.34), we may enable the FG by default in 1.34. Then users upgrade from a version which has the cache.

Just an idea: not sure if it can fix the issue.

enj replied:
> A user can turn off this feature if they're sensitive to registry unavailability enough to prepopulate the node

We should not make users choose between availability and security.

Once the feature is stable, what is the use case for turning it off?  `PullImageSecretRecheckPeriod` could have some reasonable default like 1 day to cover the majority of use cases.

If we were building this stuff from scratch today, I don't think we would ever let pods pull images that require auth without making sure that they had access to the associated image pull secret.  Just because the image is present on the node doesn't mean that a pod is allowed to use it.

cici37 commented:
/remove-sig api-machinery

fedebongio commented:
/remove-sig api-machinery

sairameshv committed:
Implement ensure secret pulled images feature

Reference: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2535-ensure-secret-pulled-images#proposal

Signed-off-by: Sai Ramesh Vanka <svanka@redhat.com>

sairameshv commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -50,12 +59,26 @@ type imageManager struct {
 var _ ImageManager = &imageManager{}
 
+const (
+	imageStateManagerPath = "/var/lib/kubelet/image_state_manager"
```
Addressed

sairameshv commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -164,17 +243,89 @@ func (m *imageManager) EnsureImageExists(ctx context.Context, pod *v1.Pod, conta
+			}
+		}()
+	}
+	byteData, err := utiljson.Marshal(m.ensureSecretPulledImages)
```
I think defining a proper structure for the image state manager's cache might be a good idea, I would work on it

sairameshv commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -164,17 +243,89 @@ func (m *imageManager) EnsureImageExists(ctx context.Context, pod *v1.Pod, conta
+	}
+	err = utiljson.Unmarshal(byteData, &m.ensureSecretPulledImages)
+	if err != nil {
+		klog.ErrorS(err, "Failed to unmarshal /var/lib/kubelet/image_state_manager file data")
```
As the worst case scenario, we end up pulling the image again for the existing images capturing the metadata again and populating the file.

sairameshv commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -164,17 +243,89 @@ func (m *imageManager) EnsureImageExists(ctx context.Context, pod *v1.Pod, conta
+	}
+	m.fsLock.Lock()
+	defer m.fsLock.Unlock()
+	err = os.WriteFile(imageStateManagerPath, byteData, 0644)
```
Sure, I would work on it

sairameshv commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -164,17 +243,89 @@ func (m *imageManager) EnsureImageExists(ctx context.Context, pod *v1.Pod, conta
 
+func (m *imageManager) storeEnsureSecretPulledImagesData() {
+	if _, err := os.Stat(imageStateManagerPath); os.IsNotExist(err) {
+		imageFile, err := os.Create(imageStateManagerPath)
```
Addressed

sairameshv commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -164,17 +243,89 @@ func (m *imageManager) EnsureImageExists(ctx context.Context, pod *v1.Pod, conta
+	if m.pullImageSecretRecheck {
+		m.lock.Lock()
+		defer m.lock.Unlock()
+		// Get the ImageRef after the Image Pull
```
Actually, we are not pulling the image here ,instead we perform the Get operation where it returns the image reference in case if the image is present locally on the machine.
This info is required just to decide if we have to perform a Pull again or not. So, we need not add mechanism to record any events related to pull right?
Am I missing something here?

mikebrow replied:
Good point @liggitt, keeping a list of in progress pull attempts is nec. to manage a disconnect between 2 and 3..

cici37 commented:
/remove-sig api-machinery

stlaz commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -220,3 +367,98 @@ func applyDefaultImageTag(image string) (string, error) {
+				klog.ErrorS(err, "Failed to hash auth", "auth", auth)
+				continue
+			}
+			m.lock.RLock()
```
you haven't released the write lock yet

liggitt commented:
moving to API review backlog for re-implementation based on 1.32 design updates in https://github.com/kubernetes/enhancements/pull/4789

liggitt commented:
/close

in favor of 1.32 implementation of design updates in https://github.com/kubernetes/enhancements/pull/4789 - https://github.com/kubernetes/kubernetes/pull/128152
